cmd/prefetchfiles-nri-plugin/main.go

```diff
@@ -0,0 +1,272 @@
+)
+
+const (
+	defaultEvents     = "RunPodSandbox"
+	defaultServerPath = "/usr/local/bin/optimizer-server"
+	defaultPersistDir = "/opt/nri/optimizer/results"
+)
+
+type PluginConfig struct {
+	Events     []string `toml:"events"`
+	ServerPath string   `toml:"server_path"`
+	PersistDir string   `toml:"persist_dir"`
+	Timeout    int      `toml:"timeout"`
+	Overwrite  bool     `toml:"overwrite"`
+}
+
+type PluginArgs struct {
+	PluginName   string
+	PluginIdx    string
+	PluginEvents string
+	Config       PluginConfig
+}
+
+type Flags struct {
+	Args *PluginArgs
+	F    []cli.Flag
+}
+
+func buildFlags(args *PluginArgs) []cli.Flag {
+	return []cli.Flag{
+		&cli.StringFlag{
+			Name:        "name",
+			Usage:       "plugin name to register to NRI",
+			Destination: &args.PluginName,
+		},
+		&cli.StringFlag{
+			Name:        "idx",
+			Usage:       "plugin index to register to NRI",
+			Destination: &args.PluginIdx,
+		},
+		&cli.StringFlag{
+			Name:        "events",
+			Value:       defaultEvents,
+			Usage:       "the events that containerd subscribes to. DO NOT CHANGE THIS.",
+			Destination: &args.PluginEvents,
+		},
+		&cli.StringFlag{
+			Name:        "server-path",
+			Value:       defaultServerPath,
+			Usage:       "the path of optimizer server binary",
+			Destination: &args.Config.ServerPath,
+		},
+		&cli.StringFlag{
+			Name:        "persist-dir",
+			Value:       defaultPersistDir,
+			Usage:       "the directory to persist accessed files list for container",
+			Destination: &args.Config.PersistDir,
+		},
+		&cli.IntFlag{
+			Name:        "timeout",
+			Value:       0,
+			Usage:       "the timeout to kill optimizer server, 0 to disable it",
+			Destination: &args.Config.Timeout,
+		},
+		&cli.BoolFlag{
+			Name:        "overwrite",
+			Usage:       "whether to overwrite the existed persistent files",
+			Destination: &args.Config.Overwrite,
+		},
+	}
+}
+
+func NewPluginFlags() *Flags {
+	var args PluginArgs
+	return &Flags{
+		Args: &args,
+		F:    buildFlags(&args),
+	}
+}
+
+type plugin struct {
+	stub stub.Stub
+	mask stub.EventMask
+}
+
+var (
+	cfg                  PluginConfig
+	log                  *logrus.Logger
+	logWriter            *syslog.Writer
+	globalFanotifyServer = make(map[string]*fanotify.Server)
+)
+
+func (p *plugin) Configure(config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	tree, err := toml.Load(config)
+	if err != nil {
+		return 0, errors.Wrap(err, "parse TOML")
+	}
+	if err := tree.Unmarshal(&cfg); err != nil {
+		return 0, err
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, errors.Wrap(err, "parse events in configuration")
+	}
+
+	log.Infof("configuration: %#v", cfg)
+
+	return p.mask, nil
+}
+
+func sendDataOverHTTP(data string, endpoint string) error {
+	url := "http://172.0.0.1:9110" + endpoint
```

Comment:
Should we POST a request to "http://172.0.0.1:9110"? This HTTP URL is used to export metrics and may not always be enabled, so we can not be fully dependent on it. I think the unix domain socket for the system controller is more reasonable.